feat: use jsonnet to pull existing alerts from mixins available

slashpai opened the pull request:
This PR adds jsonnet code to pull alerting mixins from prometheus-operator, prometheus and alertmanager. I have tried to keep jsonnet code simple as possible to manage the complexity and utilize existing things

@fpetkovski I also wanted to add ci changes for jsonnet fmt and linting checks, was not sure if we add that in same PR. I can add ci changes in same PR if thats how we do otherwise in separate PR

slashpai commented:
cc: @fpetkovski @sthaha

slashpai commented:
will require first https://github.com/rhobs/monitoring-stack-operator/pull/106 to be merged for the operator bundle tests to pass its due to name conflicts

fpetkovski commented:
I guess we also need to create these rules in the prometheus instance for the stack. Is that going to come as a follow up?

slashpai commented:
> I guess we also need to create these rules in the prometheus instance for the stack. Is that going to come as a follow up?

did you mean the prometheus deployed by prometheus operator? We should monitor the mso and other resources it created by another prometheus instance right?

fpetkovski commented:
The prometheus instances created by monitoring stacks should have alerts for themselves and for the alertmanager that belongs to that stack. Does that make sense?

slashpai commented:
Also I haven't added serviceMonitor for prometheus-operator in this PR, should we add that?

fpetkovski commented:
We can do that in another PR

slashpai commented:
> The prometheus instances created by monitoring stacks should have alerts for themselves and for the alertmanager that belongs to that stack. Does that make sense?

I can add that in separate PR. To do that we need to create separate prometheus rule manifests to match with the labels of resourceSelector labels of mso cr if my understanding is correct. To test the rules I was using same prometheus instance only by having a matching resourceSelector label.

sthaha commented on Makefile:
```diff
@@ -123,6 +131,17 @@ $(OC) oc: $(TOOLS_DIR)
+
 # Install all required tools
 .PHONY: tools
 tools: $(CONTROLLER_GEN) \
```
The jsonnet-tools need to be added here. 
And the `.github/tools` need to be updated. 
We also need .github/tools-cache/action.yaml to be updated so that the cache is correctly maintained.

slashpai replied:
I had added that initially and removed it. I thought to add once I add the ci changes also like to do lining tests and generate tests in GitHub actions. 

Do we add multiple types of commit in pr in repo since we follow conventional commits I thought to add that in another pr. But I can add in same pr if that's ok

fpetkovski replied:
It's fine to add multiple commits, and they can all have different types if you wish

sthaha commented on Makefile:
```diff
@@ -42,6 +42,14 @@ PROMQ_VERSION = v0.0.1
 OC = $(TOOLS_DIR)/oc
 OC_VERSION = v4.9.7
 
+# jsonnet related tools and dependencies
+JSONNET = $(TOOLS_DIR)/jsonnet
+JSONNETFMT = $(TOOLS_DIR)/jsonnetfmt
+JSONNET_BUNDLER = $(TOOLS_DIR)/jb
+GOJSONTOYAML = $(TOOLS_DIR)/gojsontoyaml
```
These tools need to be added to the .github/tools file : see `tools` target

sthaha commented on Makefile:
```diff
@@ -123,6 +131,17 @@ $(OC) oc: $(TOOLS_DIR)
 	}
 
+.PHONY: jsonnet-tools
+$(JSONNET) jsonnet-tools: $(TOOLS_DIR)
```
This is bit different to the installation of other tools and I think either
* we remove $(JSONNET) and any target that requries $(JSONNET*) should depend on `jsonnet-tools`
* or list all the tools that are getting installed as targets

slashpai replied:
I will remove `$(JSONNET)` and depend on `jsonnet-tools` in targets requiring this

sthaha commented on Makefile:
```diff
@@ -123,6 +131,17 @@ $(OC) oc: $(TOOLS_DIR)
+$(JSONNET) jsonnet-tools: $(TOOLS_DIR)
+	@{ \
+		set -ex ;\
+		[[ -f $(JSONNET) && -f $(JSONNETFMT) && -f $(JSONNET_BUNDLER) && -f $(GOJSONTOYAML) ]] && exit 0 ;\
```
it is better to check individually ... e.g.

```Makefile
.PHONY: jsonnet-tools
$(JSONNET) $(JSONNETFMT) $(JSONNET_BUNDLER) $(GOJSONTOYAML) jsonnet-tools: $(TOOLS_DIR)
	@{ \
			set -ex ;\
			export GOBIN=$(TOOLS_DIR) ;\
			[[ -f $(JSONNET) ]] ||  go install github.com/google/go-jsonnet/cmd/jsonnet@latest ;\
			[[ -f $(JSONNETFMT) ]] || go install github.com/google/go-jsonnet/cmd/jsonnetfmt@latest ;\
			[[ -f $(JSONNET_BUNDLER) ]] || go install github.com/jsonnet-bundler/jsonnet-bundler/cmd/jb@latest ;\
			[[ -f $(GOJSONTOYAML) ]] || go install github.com/brancz/gojsontoyaml@latest ;\
		}
```

Do you also need `-a` when installing jsonnet-bundler ? Could you please comment/doc why  that is required?

slashpai replied:
The reason I used `-a` was because per readme of [jsonnet-bundler](https://github.com/jsonnet-bundler/jsonnet-bundler#install) but it works without that when I tested I will just remove that flag

sthaha commented on Makefile:
```diff
@@ -149,6 +168,20 @@ tools: $(CONTROLLER_GEN) \
 	$(GOLANGCI_LINT) run ./... --fix
 
+.PHONY: jsonnet-fmt
+jsonnet-fmt: $(JSONNETFMT_BIN)
```
there is no target - `$(JSONNNETFMT_BIN)

slashpai replied:
good catch I forgot after I renamed the variable

sthaha commented on Makefile:
```diff
@@ -149,6 +168,20 @@ tools: $(CONTROLLER_GEN) \
 
+.PHONY: jsonnet-fmt
+jsonnet-fmt: $(JSONNETFMT_BIN)
+	find jsonnet/ -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | xargs -n 1 -- $(JSONNETFMT) -i
```
may want to consider breaking long lines for better readability 
e.g.
```Makefile

.PHONY: jsonnet-fmt
jsonnet-fmt: $(JSONNETFMT)
	find jsonnet/ -name 'vendor' -prune \
	  -o -name '*.libsonnet' -print \
		-o -name '*.jsonnet' -print \
		| xargs -n 1 -- $(JSONNETFMT) -i

```

sthaha commented on Makefile:
```diff
@@ -149,6 +168,20 @@ tools: $(CONTROLLER_GEN) \
+	find jsonnet/ -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | xargs -n 1 -- $(JSONNETFMT) -i
+
+.PHONY: generate-prometheus-rules
+generate-prometheus-rules: $(JSONNET) $(KUSTOMIZE)
```
This makes use of $(GOJSONTOYAML) so like mentioned before, the target should depend on `jsonnet-tools` or individual targets 

I think it may helpful to ensure that the env has `jq` installed 

We can also improve the readability as follows ... 

```makefile
.PHONY: check-jq
check-jq:
	jq --version > /dev/null

.PHONY: generate-prometheus-rules
generate-prometheus-rules: jsonnet-tools $(JSONNET) $(GOJSONTOYAML) $(KUSTOMIZE) check-jq
	@{ \
		set -e ;\
		cd jsonnet && jb install && cd - ;\
		\
		for dir in jsonnet/components/*/; do \
			component=$$(basename $$dir) ;\
			echo "Generating prometheusrule file for $$component" ;\
			$(JSONNET) -J jsonnet/vendor $$dir/main.jsonnet \
			| jq .rule  \
			| $(GOJSONTOYAML) > deploy/operator/monitoring-stack-operator-$$component-rules.yaml ;\
			\
			cd deploy/operator && \
				$(KUSTOMIZE) edit add resource "monitoring-stack-operator-$$component-rules.yaml" && \
			cd - >/dev/null ;\
		done \
	}
```

slashpai commented:
I updated the `Makefile` per review comments and made it a little better to manage targets better. I also added ci changes, at this point only `jsonnetfmt` and `jsonnet-lint` is added which needs `jsonnet-bundler` as well. So only these tools I have added in tools cache and actions. I am figuring out how we can test jsonnet-generate since we don't generate alerts yamls directly instead prometheuerule files with the help of `gojsontoyaml`

slashpai commented:
I temporarily removed jobs from pr-checks.yaml other than tools-cache. Trying to get this pass since it was failing with oc installation

slashpai commented:
@sthaha @fpetkovski : Can you review again and see if we are good to merge now?

slashpai commented on .github/tools:
```diff
@@ -1,6 +1,9 @@
 controller-gen v0.7.0
 kustomize v3.9.4
-oc v4.9.7
+oc v4.8.11
```
Had to do this since mirror was trying to install this version and tools cache validation was failing without this

```bash
+ rm -rf /tmp/tmp.sFQfPvVtFW
+ [[ -f /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin/oc ]]
+ OS=linux
+ ARCH=amd64
+ curl -sSLo /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin/oc https://mirror.openshift.com/pub/openshift-v4/amd64/clients/oc/latest/linux/oc.tar.gz
+ tar -xf /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin/oc -C /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin
+ rm -f /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin/README.md
+ /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin/oc version
Client Version: 4.8.11
+ version=v4.9.7
+ /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin/oc version
+ grep -q 4.9.7
make: *** [Makefile:132: /home/runner/work/monitoring-stack-operator/monitoring-stack-operator/tmp/bin/oc] Error 1
```

sthaha replied:
The right way to do this is to edit the makefile

slashpai commented:
I created separate commits since changes are different, we might not want to squash them :)

sthaha commented on Makefile:
```diff
@@ -123,14 +141,60 @@ $(OC) oc: $(TOOLS_DIR)
+.PHONY: jsonnetfmt
+$(JSONNETFMT) jsonnetfmt: $(TOOLS_DIR)
+	@{ \
+		set -ex ;\
+		export GOBIN=$(TOOLS_DIR) ;\
+	  [[ -f $(JSONNETFMT) ]] || go install github.com/google/go-jsonnet/cmd/jsonnetfmt@latest ;\
```
these one liners don't actually need `@{ ... }` hack and can be simplified as follows
```
target: deps 
  GOBIN=$(TOOLS_DIR) go install .... 
```

`[[ -f ]]` is superfluous since `go install` will take care of that.

sthaha commented on Makefile:
```diff
@@ -123,14 +141,60 @@ $(OC) oc: $(TOOLS_DIR)
+.PHONY: jsonnetfmt
+$(JSONNETFMT) jsonnetfmt: $(TOOLS_DIR)
+	@{ \
+		set -ex ;\
+		export GOBIN=$(TOOLS_DIR) ;\
```
nit: alignment

sthaha commented on Makefile:
```diff
@@ -149,6 +216,40 @@ tools: $(CONTROLLER_GEN) \
+jsonnet-lint: $(JSONNETLINT) jsonnet-vendor
+	find jsonnet/ -name 'vendor' -prune \
+	  -o -name '*.libsonnet' -print \
+		-o -name '*.jsonnet' -print \
```
nit: alignment

sthaha commented on Makefile:
```diff
@@ -40,7 +40,25 @@ PROMQ_VERSION = v0.0.1
+JSONNETFMT = $(TOOLS_DIR)/jsonnetfmt
+JSONNETFMT_VERSION = v0.17.0
+
+JSONNETLINT = $(TOOLS_DIR)/jsonnet-lint
```
`JSONNET_LINT` to match `jsonnet-lint`

sthaha commented on Makefile:
```diff
@@ -123,14 +141,60 @@ $(OC) oc: $(TOOLS_DIR)
 
+.PHONY: jsonnet
+$(JSONNET) jsonnet: $(TOOLS_DIR)
+	@{ \
+		set -ex ;\
+		export GOBIN=$(TOOLS_DIR) ;\
+		[[ -f $(JSONNET) ]] ||  go install github.com/google/go-jsonnet/cmd/jsonnet@latest ;\
```
this oneliner can actually be simplied as 
```
GOBIN=$(TOOLS_DIR) go install ....
```

sthaha commented on Makefile:
```diff
@@ -123,14 +141,60 @@ $(OC) oc: $(TOOLS_DIR)
+	}
+
+.PHONY: jsonnetbundler
+$(JSONNET_BUNDLER) jsonnetbundler: $(TOOLS_DIR)
```
jsonnetbundler -> jb to match the name of the binary

sthaha commented on Makefile:
```diff
@@ -123,14 +141,60 @@ $(OC) oc: $(TOOLS_DIR)
+		set -ex ;\
+		export GOBIN=$(TOOLS_DIR) ;\
+		[[ -f $(JSONNET_BUNDLER) ]] || go install github.com/jsonnet-bundler/jsonnet-bundler/cmd/jb@latest ;\
+		[[ -f $(GOJSONTOYAML) ]] || go install github.com/brancz/gojsontoyaml@latest ;\
```
remove gojsontoyaml?

sthaha commented on .github/workflows/pr-checks.yaml:
```diff
@@ -34,6 +34,28 @@ jobs:
+      - name: Use tools cache
+        uses: ./.github/tools-cache
+
+      - run: make jsonnet-fmt && git diff --exit-code
```
nice!

sthaha commented on Makefile:
```diff
@@ -303,3 +404,6 @@ kind-cluster: $(OPERATOR_SDK)
 	kubectl create -k deploy/dependencies
 
+.PHONY: clean
+clean:
```
Do not delete `tmp/`
Let's only cleanup what gets generated.

sthaha commented on Makefile:
```diff
@@ -149,6 +216,40 @@ tools: $(CONTROLLER_GEN) \
+		component=$$(basename $$dir) ;\
+		echo "Generating prometheusrule file for $$component" ;\
+		$(JSONNET) -J $(JSONNET_VENDOR) $$dir/main.jsonnet \
+		| jq .rule \
```
How do we ensure that this target is called? 
Should this be added to `generate` target?

sthaha commented on jsonnet/components/prometheus/prometheus-rules.jsonnet:
```diff
@@ -0,0 +1,17 @@
+/*Cannot use https://github.com/prometheus/prometheus/blob/main/documentation/prometheus-mixin/mixin.libsonnet since it generates yaml with double quotes wrapped*/
```
nit: break  long lines into two 
FYI `//` is supported by jsonnet

sthaha commented on jsonnet/components/prometheus/prometheus-rules.jsonnet:
```diff
@@ -0,0 +1,17 @@
+/*Cannot use https://github.com/prometheus/prometheus/blob/main/documentation/prometheus-mixin/mixin.libsonnet since it generates yaml with double quotes wrapped*/
+
+local rules = (
+  import 'github.com/prometheus/prometheus/documentation/prometheus-mixin/mixin.libsonnet'
```
doesn't this contradict the comment above?

slashpai replied:
no but I can modify the sentence. what I meant was I cannot directly use import, I do extract `.prometheusAlerts` from the import

sthaha replied:
aha! it would be nice if you could clarify what you meant in the comment itself 👼

sthaha commented on Makefile:
```diff
@@ -40,7 +40,25 @@ PROMQ_VERSION = v0.0.1
+JSONNETFMT_VERSION = v0.17.0
+
+JSONNETLINT = $(TOOLS_DIR)/jsonnet-lint
+JSONNETLINT_VERSION = v0.17.0
```
```suggestion
JSONNET_LINT_VERSION = v0.17.0
```

sthaha commented on Makefile:
```diff
@@ -40,7 +40,25 @@ PROMQ_VERSION = v0.0.1
+JSONNETLINT = $(TOOLS_DIR)/jsonnet-lint
+JSONNETLINT_VERSION = v0.17.0
+
+JSONNET_BUNDLER = $(TOOLS_DIR)/jb
```
```suggestion
JB = $(TOOLS_DIR)/jb
```

sthaha commented on Makefile:
```diff
@@ -40,7 +40,25 @@ PROMQ_VERSION = v0.0.1
+JSONNETLINT_VERSION = v0.17.0
+
+JSONNET_BUNDLER = $(TOOLS_DIR)/jb
+JSONNET_BUNDLER_VERSION = v0.4.0
```
```suggestion
JB_VERSION = v0.4.0
```

sthaha commented on jsonnet/common.libsonnet:
```diff
@@ -0,0 +1,17 @@
+          labels: labels,
+          name: name,
+        },
+        spec:
```
```suggestion
        spec: rules,
```

sthaha commented on Makefile:
```diff
@@ -149,6 +216,40 @@ tools: $(CONTROLLER_GEN) \
+check-jq:
+	jq --version > /dev/null
+
+jsonnet-vendor: $(JSONNET_BUNDLER)
```
this forces the bundles to be downloaded every time. is this needed, if so, could you please add a comment why this is needed?

sthaha commented on Makefile:
```diff
@@ -149,6 +216,40 @@ tools: $(CONTROLLER_GEN) \
+	jq --version > /dev/null
+
+jsonnet-vendor: $(JSONNET_BUNDLER)
+	rm -rf jsonnet/vendor
```
```suggestion
	
```

sthaha commented on Makefile:
```diff
@@ -123,32 +145,99 @@ $(OC) oc: $(TOOLS_DIR)
 
+.PHONY: jsonnet
+$(JSONNET) jsonnet: $(TOOLS_DIR)
+		GOBIN=$(TOOLS_DIR) go install github.com/google/go-jsonnet/cmd/jsonnet@latest
```
```suggestion
	GOBIN=$(TOOLS_DIR) go install github.com/google/go-jsonnet/cmd/jsonnet@latest
```

sthaha commented on Makefile:
```diff
@@ -123,32 +145,99 @@ $(OC) oc: $(TOOLS_DIR)
+
+.PHONY: jsonnetfmt
+$(JSONNETFMT) jsonnetfmt: $(TOOLS_DIR)
+		GOBIN=$(TOOLS_DIR)  go install github.com/google/go-jsonnet/cmd/jsonnetfmt@latest
```
```suggestion
	GOBIN=$(TOOLS_DIR)  go install github.com/google/go-jsonnet/cmd/jsonnetfmt@latest
```

sthaha commented on .github/workflows/pr-checks.yaml:
```diff
@@ -34,6 +34,28 @@ jobs:
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Use tools cache
```
we should include the `go-cache` before tools cache since go mods will be reused.

sthaha reviewed (approved):
Good stuff! @slashpai

slashpai commented:
Requesting review from @fpetkovski if we need anymore changes :)

fpetkovski reviewed (approved):
Nice!

slashpai commented:
I had moved all lint checks under a single workflow job. I see all checks passed but waiting for golang lint check results to enable merge, I cannot see job url displayed

fpetkovski commented:
It seems that this job was added as required for merging to main. Removing it from the settings seems to have done the trick
